File: solutions/java/proverb/src/main/java/Proverb.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

record Proverb(String[] words) {
    String recite() {
        return IntStream.range(0, words.length)
                .mapToObj(this::line)
                .collect(Collectors.joining("\n"));
    }

    private String line(int i) {
        return i == words.length - 1
            ? "And all for the want of a " + words[0] + "."
            : "For want of a " + words[i] + " the " + words[i + 1] + " was lost.";
    }
}

File: solutions/python/proverb/proverb.py
def proverb(*things: str, qualifier: str | None) -> list[str]:
    last_thing = len(things) - 1
    return [
        f"For want of a {things[i]} the {things[i + 1]} was lost." if i < last_thing else
        f"And all for the want of a {qualifier + ' ' if qualifier else ''}{things[0]}."
        for i in range(len(things))
    ]
